### PR TITLE
Implementation of Select2 wrapper into Popover

### PR DESCRIPTION
**Pull request by NejcZdovc**
Implementation of select2 wrapper into the Popover, tab Keypress (#45).

@fjozsef and @mondalaci please review the code and give a feedback. Thank you

### CODE REVIEW

**Comment by fjozsef** [comment]
Could you move the changes of the keypress-tab to another PR?
Otherwise the wrapper LGTM.

**Comment by fjozsef** [comment]
The long press action needs the be aligned vertically center.
Could you rename the commit to "Use select2 in keypress-tab"?

**Comment by NejcZdovc** [comment]
I will check on the css problems today.

I wanted to use this pull request for implementation of whole popover, so it's maybe better to rename to "Use select2 in popover".

**Comment by fjozsef** [comment]
Okay, I think it's good start. Let me know if you finished the replacements in the whole popover.

**Comment by NejcZdovc** [comment]
Fixed vertical alignment. Problem with the layout is because text is too long. We can resize select2 or provide shorter text for this two options or reduce font size.

**Comment by NejcZdovc** [comment]
I analysed other tabs and I will need to add output to the wrapper so that other functionalities can work based on the current selection. Will work on this.

**Comment by NejcZdovc** [comment]
@fjozsef Select2 is now added to all tabs in popup. Please review commits. What should we do with the text that is too long in Keypress tab?

File: src/components/popover/tab/layer-tab.component.ts [hunk 1]
```diff
@@ -1,32 +1,58 @@
...
-            <option [ngValue]="0"> Mod </option>
-            <option [ngValue]="1"> Fn </option>
-            <option [ngValue]="2"> Mouse </option>
-        </select>
-        <span>
-        layer by holding this key.
+        <select2 [data]="layerData" (valueChanged)="layerChanged($event)"></select2>
+        <span [ngSwitch]="toggle">
+             <template ngSwitchWhen="true">layer by pressing this key.</template>
+             <template ngSwitchDefault="false">layer by holding this key.</template>
```
Comment (fjozsef) [inline src/components/popover/tab/layer-tab.component.ts]: It is default false. Maybe the toggle doesn't change. Can you look into that?

File: src/components/popover/tab/layer-tab.component.ts [hunk 2]
```diff
@@ -45,5 +71,12 @@ export class LayerTabComponent implements OnInit, KeyActionSaver {
         keyAction.layer = this.layer;
         return keyAction;
     }
+    
+    private toggleChanged(event) {
```
Comment (fjozsef) [inline src/components/popover/tab/layer-tab.component.ts]: This function wasn't called when I tested.

File: src/components/popover/tab/layer-tab.component.ts [hunk 3]
```diff
@@ -45,5 +71,12 @@ export class LayerTabComponent implements OnInit, KeyActionSaver {
...
+        this.toggle = event.value;
+    }
 
+    private layerChanged(event) {
```
Comment (fjozsef) [inline src/components/popover/tab/layer-tab.component.ts]: This neither.

**Comment by fjozsef** [comment]
I've checked the other change handling functions and they haven't been called either. Maybe it is bug in the wrapper?

**Comment by NejcZdovc** [comment]
Which version of the wrapper do you have installed? Output was added in version 0.3.1 of the wrapper.

**Comment by fjozsef** [comment]
I thought I did, but now I have reinstalled all the dependencies and recompiled and it works. Sorry, I guess I skipped something.

File: src/components/popover/tab/keymap-tab.component.ts [hunk 4]
```diff
@@ -6,17 +6,17 @@ import {SvgKeyboardComponent} from '../../svg-keyboard.component';
 import {KeyActionSaver} from '../key-action-saver';
 import {SwitchKeymapAction} from '../../../../config-serializer/config-items/SwitchKeymapAction';
 
+import {OptionData} from "ng2-select2/dist/select2";
```
Comment (fjozsef) [inline src/components/popover/tab/keymap-tab.component.ts]: Use ' instead of ". The linter will show you where you missed.

**Comment by fjozsef** [comment]
1. Run the tslinter, correct the warning you have (note: there were some warnings before your change).
2. Squash into one commit.

**Comment by NejcZdovc** [comment]
This is something that I wanted to talk about it with you. I run linter and there is no errors. I know there were some, so I need to figure it out why it's not working.
image

**Comment by fjozsef** [comment]
Try 'npm run tslint'

**Comment by NejcZdovc** [comment]
There is a problem with paths. `tslint 'src/**/*.ts'` this is working, so there should be different implementation for multiple paths.

**Comment by fjozsef** [comment]
Are you using Windows, right? Could you send another PR  if you managed to solve that issue? I can test the Linux side.

**Comment by NejcZdovc** [comment]
Sadly Windows, but not for long :) Will try to solve it.

**Comment by NejcZdovc** [comment]
Commits squashed and linter warnings fixed.

File: src/components/popover/tab/macro-tab.component.ts [hunk 5]
```diff
@@ -42,6 +42,22 @@ export class MacroTabComponent implements OnInit, KeyActionSaver {
 
     ngOnInit() {
         this.macros = this.uhkConfigurationService.getUhkConfiguration().macros.elements;
+
+        this.macrosOptions.push({
+            id: '-1',
+            text: 'Select macro'
+        });
+
+        for (let value of this.macros) {
```
Comment (mondalaci) [inline src/components/popover/tab/macro-tab.component.ts]: @NejcZdovc Wouldn't this be cleaner to implement in a functional manner with `Array.map()`?

Comment (NejcZdovc) [inline src/components/popover/tab/macro-tab.component.ts]: Fixed in commit b4a1625

**Comment by NejcZdovc** [comment]
@fjozsef and @mondalaci please if everything is in order merge this to the master branch, so there will be no new and unnecessary conflicts.

**Comment by fjozsef** [comment]
@NejcZdovc Sorry for the latency. I will try to review it today, but I didn't want to merge without trying out the latest commit. 

PS: I don't work on this project in full time, so sometimes you have to wait, but I try the review the PR-s as soon as I am not busy (in 3 days, or in 5-7 days in the worst case).

**Pull request merged**